runtime/near-vm-logic/src/logic.rs

```diff
@@ -943,14 +944,87 @@ impl<'a> VMLogic<'a> {
-        self.gas_counter.pay_per_byte(keccak512_byte, value.len() as u64)?;
+        self.gas_counter.pay_per(keccak512_byte, value.len() as u64)?;
 
         use sha3::Digest;
 
         let value_hash = sha3::Keccak512::digest(&value);
         self.internal_write_register(register_id, value_hash.as_slice().to_vec())
     }
 
+    /// Hashes the given value using RIPEMD-160 and returns it into `register_id`.
+    ///
+    /// # Errors
+    ///
+    /// If `value_len + value_ptr` points outside the memory or the registers use more memory than
+    /// the limit with `MemoryAccessViolation`.
+    ///
+    /// # Cost
+    ///
+    ///  Where `message_blocks` is `(value_len + 9).div_ceil(64)`.
+    ///
+    /// `base + write_register_base + write_register_byte * num_bytes + ripemd160_base + ripemd160_block * message_blocks`
+    pub fn ripemd160(&mut self, value_len: u64, value_ptr: u64, register_id: u64) -> Result<()> {
+        use num_integer::Integer;
+
+        self.gas_counter.pay_base(ripemd160_base)?;
+        let value = self.get_vec_from_memory_or_register(value_ptr, value_len)?;
+
+        let message_blocks = (value_len + 9).div_ceil(&64);
+
+        self.gas_counter.pay_per(ripemd160_block, message_blocks)?;
+
+        use ripemd160::Digest;
+
+        let value_hash = ripemd160::Ripemd160::digest(&value);
+        self.internal_write_register(register_id, value_hash.as_slice().to_vec())
+    }
+
+    /// Recovers an ECDSA signer address and returns it into `register_id`.
+    ///
+    /// Takes in an additional flag to check for malleability of the signature
+    /// which is generally only ideal for transactions.
+    ///
+    /// Returns a bool indicating success or failure as a `u64`.
+    ///
+    /// # Errors
+    ///
+    /// * If `hash_ptr`, `r_ptr`, or `s_ptr` point outside the memory or the registers use more
+    ///   memory than the limit, then returns `MemoryAccessViolation`.
+    ///
+    /// # Cost
+    ///
+    /// `base + write_register_base + write_register_byte * 20 + ecrecover_base`
+    pub fn ecrecover(
+        &mut self,
+        hash_ptr: u64,
+        sig_ptr: u64,
+        malleability_flag: u8,
+        register_id: u64,
+    ) -> Result<u64> {
+        self.gas_counter.pay_base(ecrecover_base)?;
+
+        let mut signature_bytes = [0u8; 65];
+        self.memory_get_into(sig_ptr, &mut signature_bytes)?;
+        let signature = Secp256K1Signature::from(signature_bytes);
+
+        let mut hash_bytes = [0u8; 32];
+        self.memory_get_into(hash_ptr, &mut hash_bytes)?;
+
+        if malleability_flag == 1 {
+            if !signature.check_signature_values() {
+                return Ok(false as u64);
+            }
+        }
+
+        if let Ok(pk) = signature.recover(hash_bytes) {
+            self.internal_write_register(register_id, pk.as_ref().to_vec())?;
+            return Ok(true as u64);
+        };
+
+        return Ok(false as u64);
+    }
```

Comment:
There are multiple issues:
1. Unless there is a good reason, please use register API (`get_vec_from_memory_or_register`) to get the inputs.
2. Because different protocols encode the `v` value differently, pass it as a separate argument rather than as a part of the signature. The signature should be 64 bytes.
3. The value of `v` should be in the range [0, 3]. This is what most secp256k1 implementations expect.
4. Now, it can be combined with `malleability_flag` into a single argument, `flags`. Bits 0 and 1 would be `v`, and bit 2 would be the malleability flag.
5. Also, as there is no `u8` type in WASM, this argument's type should be `u32`.
6. The malleability flag should only control the malleability check. The `check_signature_values` performs other checks as well. However, they seems to be redundant, because `signature.recover` performs them as well.

Reply:
1. I had done so to prevent another vec initialisation, as recommended by @matklad. I'll change this back though.
2. Good call
3. Right I'll add this in.
4. I'm unsure if that is a good idea as it isn't exactly explicit enough without reading the documentation that the `v` is in there and I don't exactly how they are directly related in that sense. Let me know if you still want to do that.
5. That has been changed.
6. I'll take a look at that and remove the ones that are already done.

Reply:
>Unless there is a good reason, please use register API (get_vec_from_memory_or_register) to get the inputs.

I believe there's good reason to not do this here, but this isn't a blocker. 

In this case, we have inputs of _fixed length_: the caller doesn't provide the len, its hard-coded to 64 at the call-site. So, `get_vec_from_memory_or_register` is actually guaranteed to fetch the vec from memory, not from register (that function uses `len` to decide between memory and register). By reading directly into fixed-size buffer we gain two benefits: better correctness (we track the size at compile time) and better performance (no need to allocate a vec). 

However, as it's impossible to use registers with this API, the choice here doesn't affect the ABI and so can be revised later. 

To be clear, using `get_vec_from_memory_or_register` is not wrong, its just a bit suboptimal and confusing for fixed-sized inputs.

Reply:
I do agree with this @matklad. If @abacabadabacaba is good with that, then I'll add that back in.